python/pysrc/kaskada/windows.py

```diff
@@ -55,9 +62,73 @@ def yearly() -> Since:
         return Since(predicate=lambda domain: Timestream._call("yearly", domain))
 
 
+@dataclass(frozen=True)
+class Tumbling(Window):
+    """Window producing a single value at the time of a predicate.
+
+    "Tumbling" windows are a series of non-overlapping windows producing a single value
+    at the time the predicate evaluates to true. Each window is inclusive of the time
+    it starts at and exclusive of the end time.
+
+    Args:
+        predicate: the condition used to determine when the window resets.
+
+    Returns:
+        Window for aggregating cumulative values since the predicate.
+    """
+
+    #: The boolean Timestream to use as predicate for the window.
+    #: Each time the predicate evaluates to true the window will be cleared.
+    #:
+    #: The predicate may be a callable which returns the boolean Timestream, in
+    #: which case it is applied to the Timestream being aggregated.
+    predicate: Timestream | Callable[..., Timestream] | bool
+
+    @staticmethod
+    def minutely() -> Tumbling:
+        """Return a tumbling window that resets at the start of each minute."""
+        return Tumbling(predicate=lambda domain: Timestream._call("minutely", domain))
+
+    @staticmethod
+    def hourly() -> Tumbling:
+        """Return a tumbling window that resets at the start of each hour."""
+        return Tumbling(predicate=lambda domain: Timestream._call("hourly", domain))
+
+    @staticmethod
+    def daily() -> Tumbling:
+        """Return a tumbling window that resets at the start of each day."""
+        return Tumbling(predicate=lambda domain: Timestream._call("daily", domain))
+
+    @staticmethod
+    def monthly() -> Tumbling:
+        """Return a tumbling window that resets at the start of each month."""
+        return Tumbling(predicate=lambda domain: Timestream._call("monthly", domain))
+
+    @staticmethod
+    def yearly() -> Tumbling:
+        """Return a tumbling window that resets at the start of each year."""
+        return Tumbling(predicate=lambda domain: Timestream._call("yearly", domain))
+
+
 @dataclass(frozen=True)
 class Sliding(Window):
-    """Window for the last `duration` intervals of some `predicate`."""
+    """Overlapping windows producing the latest cumulative value since a predicate.
+
+    "Sliding" windows are a series of overlapping windows producing cumulative values
+    since each time the predicate evaluates to true. Each window is inclusive of the time
+    it starts at and exclusive of the end time.
+
+    If multiple windows exist, the value produced at any given time will be the cumulative
+    value in the most recently created window.
+
+    Args:
+        duration: the number of active windows at any given time.
+        predicate: the condition used to determine when the oldest window ends and a new
+          window starts.
+
+    Returns:
+        Overlapping windows for aggregating cumulative values since the predicate.
```

Comment:
Since the N-th predicate evaluation to true?

I wonder if these would be easier if we described this as a "boundary" or "segment" or something. Basically, window boundaries exist when the predicate evaluates to true. Since windows produce cumulative values since the most recent boundary. Sliding windows produce cumulative values since the Nth most recent boundary. Tumbling windows produce the value since the last boundary at each boundary.

Reply:
Just going to leave it as `overlapping windows for aggregation values` -- I highly anticipate we'll change the parameters for these